### PR TITLE
Support for chunked analysis

### PR DESCRIPTION
**Pull request by rawler**
This PR enables an application to do "chunked" analysis of long streams. When analyzing multiple hours of audio, analyzing it in chunks allows usage of multi-core to accelerate the analysis, for both decoding and the analysis itself.

### CODE REVIEW

File: src/ebur128.rs [hunk 1]
```diff
@@ -709,6 +717,46 @@ impl EbuR128 {
         self.add_frames(crate::Planar::new(frames)?)
     }
 
+    /// Add interleaved frames to be processed.
```
Comment (sdroege) [inline src/ebur128.rs]: Can you add some documentation here that explains what this function actually does and how it differs from `add_frames()`, together with an explanation when/why one might want to do that?

File: src/ebur128.rs [hunk 2]
```diff
@@ -924,6 +972,49 @@ impl EbuR128 {
 
         Ok(self.block_energy_history.relative_threshold())
     }
+
+    pub fn try_merge(&mut self, other: Self) -> Result<&mut Self, ParameterMismatch> {
```
Comment (sdroege) [inline src/ebur128.rs]: Same for this

File: src/ebur128.rs [hunk 3]
```diff
@@ -924,6 +972,49 @@ impl EbuR128 {
...
+            .merge(other.block_energy_history)?;
+        crate::utils::merge_slice_max(&mut self.sample_peak, &other.sample_peak);
+        crate::utils::merge_slice_max(&mut self.true_peak, &other.true_peak);
+
+        Ok(self)
+    }
+}
+
+#[derive(Clone, Debug, Copy, PartialEq, Eq)]
+pub enum ParameterMismatch {
```
Comment (sdroege) [inline src/ebur128.rs]: Should implement the `std::error::Error` trait, but maybe we can also continue using the old `Error` enum here?

File: src/ebur128.rs (hunk 2 above)
Comment (rawler) [reply src/ebur128.rs]: Hmm. The idea was to be able to divide a long signal into chunks, to enable parallel analysis over the chunks. This method was intended to let "Iterator::fold" combine all the results at the end. However, when fixing this PR myself, I realized there is already `loudness_global_multiple` for getting the global loudness that I want. That means this entire merge-functionality is strictly not needed. (I also need global true-peak, but that's easily solved without explicit support in this library). I'll remove this, and simply fix the seeding-docs as requested.

**Comment by rawler** [comment]
Sorry for missing `loudness_global_multiple` in the first iteration of this. Turns out, all required for bit-exact parallel analysis is the seeding, to solve initialization of filter-states.

**Comment by rawler** [comment]
If we are willing to add a (feature-gated) dependency on rayon, I could perhaps add some "ParallelAnalyzer" implementation, that expects a stream of `impl Samples`, and through Rayon automatically chunks and analyzes the stream in parallel. I'm going to write such implementation for our application anyways. Let me know if that's desirable.

**Comment by sdroege** [comment]
That sounds interesting but I'm not sure how that would like API-wise. What did you have in mind for the API?

File: src/ebur128.rs [hunk 4]
```diff
@@ -719,6 +779,13 @@ impl EbuR128 {
     }
 
     /// Get global integrated loudness in LUFS across multiple instances.
+    ///
+    /// This can be used to allow parallel iteration of long signals, assuming some care is taken:
+    ///  1. Divide input-signal up in "chunks" of even 100ms samples. Make chunks overlap by 400ms, for example (0-10s, 9.6-20s, 19.6-30s, ...)
+    ///  2. For each chunk, create a new instance of EbuR128, then in parallel:
```
Comment (sdroege) [inline src/ebur128.rs]:
```suggestion
    ///  2. For each chunk, create a new instance of `EbuR128`, then in parallel:
```

File: src/ebur128.rs [hunk 5]
```diff
@@ -719,6 +779,13 @@ impl EbuR128 {
...
+    ///     1. Feed the first 100ms of the chunk through seed_frames_* function. This is sufficient to make filter-states in each instance what they would have been if a single analyzer would have reached this point.
```
Comment (sdroege) [inline src/ebur128.rs]:
```suggestion
    ///     1. Feed the first 100ms of the chunk through `seed_frames_*` function. This is sufficient to make filter-states in each instance what they would have been if a single analyzer would have reached this point.
```

Comment (sdroege) [reply src/ebur128.rs]: Also that shouldn't be the first 100ms of each chunk, but the last 100ms of the previous chunk or not?

File: src/ebur128.rs [hunk 6]
```diff
@@ -719,6 +779,13 @@ impl EbuR128 {
...
+    ///     2. Process the remaining samples of each chunk through the analyzer
+    ///  3. Call loudness_global_multiple over all the chunks to get the global loudness
```
Comment (sdroege) [inline src/ebur128.rs]:
```suggestion
    ///  3. Call [`loudness_global_multiple`] over all the chunks to get the global loudness
```

Does this link and the others in `seed_samples()` work like this btw?

File: src/ebur128.rs [hunk 7]
```diff
@@ -935,6 +1002,16 @@ mod tests {
     #[cfg(feature = "c-tests")]
     use quickcheck_macros::quickcheck;
 
+    fn f64_max(mut values: impl Iterator<Item = f64>) -> Option<f64> {
+        let mut v = values.next();
+        for candidate in values {
+            if candidate > v.unwrap() {
```
Comment (sdroege) [inline src/ebur128.rs]: This seems like a suboptimal implementation as that can easily produce all the panic code, unwinding code, etc.

Maybe this should be

```rust
let mut v = values.next()?;
for candidate in values {
    if candidate > v {
        v = candidate;
    }
}

Some(v)
```

File: src/filter.rs [hunk 8]
```diff
@@ -258,6 +253,64 @@ impl Filter {
                     }
                 }
             }
+
+            if let Some(ref mut tp) = self.tp {
+                assert!(self.true_peak.len() == self.channels as usize);
+                tp.check_true_peak(src, &mut *self.true_peak);
+            }
```
Comment (sdroege) [inline src/filter.rs:260]: Why did you move this after the filter now, and outside the `ftz`?

File: src/true_peak.rs [hunk 9]
```diff
@@ -164,11 +164,16 @@ impl TruePeak {
...
         &mut self,
-        src: &S,
+        src: S,
         peaks: &mut [f64],
     ) {
         self.interp.check_true_peak(src, peaks)
     }
+
+    pub fn seed<'a, T: Sample + 'a, S: crate::Samples<'a, T>>(&mut self, src: S) {
+        let mut true_peaks = vec![0.0; src.channels()];
```
Comment (sdroege) [inline src/true_peak.rs]: Sad heap allocation :) Maybe use a `SmallVec`here with a reasonable number of channels?

**Comment by sdroege** [comment]
Thanks, looks mostly good to me :)

> That sounds interesting but I'm not sure how that would like API-wise. What did you have in mind for the API?

I would still be interested in this though!

File: src/ebur128.rs (hunk 5 above)
Comment (rawler) [reply src/ebur128.rs]: EDIT: Changed my answer after re-reading your comment.

That depends on how you view a "chunk", as separate overlap and content, or simply opaque chunks, chosen to overlap 400ms. The easiest way to think about it is probably to view them as opaque chunks, of N+400ms size, where the 400ms are chosen to be overlapping.

Then technically, the difference is whether the chunks will be `[0..10, 9.6..20]` or `[0..10.4, 10..20.4]`, but in either case, the first 400 ms will be "the overlap with previous chunk", and both models would work equally fine. Maybe I can make it more clear in a parenthesis though.

File: src/ebur128.rs (hunk 6 above)
Comment (rawler) [reply src/ebur128.rs]: Will have to check

File: src/filter.rs (hunk 8 above)
Comment (rawler) [reply src/filter.rs:260]: check_true_peak will now consume "src", so either something must be done to allow cheap cloning of src (before, src was &S, so the reference where `Copy`), or this rearrangement is required.

I don't think it's moved outside the ftz though?

File: src/ebur128.rs (hunk 6 above)
Comment (rawler) [reply src/ebur128.rs]: It does, with a slight modification. In particular, all the other links in this doc, on the form ``[`EbuR128::add_frames`]`` wil work without `(struct.EbuR128.html#method.add_frames_i16)`

File: src/true_peak.rs (hunk 9 above)
Comment (rawler) [reply src/true_peak.rs]: Sure! Though, this is going to be done at most once per chunk of data analyzed. Probably not going to matter,

**Comment by rawler** [comment]
> That sounds interesting but I'm not sure how that would like API-wise. What did you have in mind for the API?

Sorry, I spent some days over last week building this, just to realize that decoding the audio itself is ~40% of the CPU spent in our application. So, analyzing chunks in parallel didn't really improve performance more than simply decoding in a separate thread does. If we want to chunk, we'll going to have to create separate decoders + analyzers for each chunk, and that's not something that I can see how to build in a generic way.

FWIW though; what I built and then dropped was roughly `EbuR128::analyze_stream<E>(self, samples: impl Iterator<Result<impl AsRef<[f32]>, E>>) -> Result<AnalysisResult, E>`. I could cleanup and bring the implementation here, if you want to write tests and maintain it, but my _guess_ is that others will hit the same "decoding" bottleneck that I did and this will bring more pain than gain.

**Comment by sdroege** [comment]
I guess for that to be useful you need a really cheap compression algorithm (say, MPEG-1 Layer 2 or even cheaper like ADPCM, or WavPack) or work directly on uncompressed audio. I can see how decoding of a more sophisticated codec is strictly more expensive than measuring the loudness, even if the resampling for true peak detection is needed.

I'm not sure, it seems like a nice API to have but also doesn't seem useful in the majority of cases. Maybe you could provide it as an example, not as API, and we just put it into the examples subdirectory? :)

File: src/filter.rs (hunk 8 above)
Comment (sdroege) [reply src/filter.rs:260]: No, I just failed reading the diff correctly. This shouldn't matter then

**Review by sdroege — APPROVED** [review]
